### PR TITLE
[RFC] Support for CTE queries under PostgreSQL

### PR DESCRIPTION
**Pull request by warrenseymour**
- Adds the `.with(alias, table)` method to query builders when using
  the 'postgres' flavour

I'd like to get feedback on whether this feature as I've written it is of interest and worthy of a merge before I proceed any further.

### CODE REVIEW

File: test/postgres.test.coffee [hunk 1]
```diff
@@ -79,6 +89,11 @@ test['Postgres flavour'] =
       toString: ->
         assert.same @del.toString(), 'DELETE FROM table WHERE (field = 1) RETURNING field'
 
+    '>> from(table).where(field = 1).with(alias, table)':
+      beforeEach: -> @del.from('table').where('field = 1').with('alias', squel.select().from('table'))
+      toString: ->
```
Comment (hiddentao) [inline test/postgres.test.coffee]: Please add a `toParam()` test too, and the `with` query should also use params, e.g. `select().from('table').where('a = ?', 5)`

File: test/postgres.test.coffee [hunk 2]
```diff
@@ -66,6 +71,11 @@ test['Postgres flavour'] =
       toString: ->
         assert.same @upd.toString(), 'UPDATE table SET field = 1 FROM table2'
 
+    '>> table(table).set(field, 1).with(alias, table)':
+      beforeEach: -> @upd.table('table').set('field', 1).with('alias', squel.select().from('table'))
+      toString: ->
```
Comment (hiddentao) [inline test/postgres.test.coffee]: Please add a `toParam()` test too, and the `with` query should also use params, e.g. `select().from('table').where('a = ?', 5)`

File: test/postgres.test.coffee [hunk 3]
```diff
@@ -48,6 +48,11 @@ test['Postgres flavour'] =
       toString: ->
         assert.same @inst.toString(), 'INSERT INTO table (field) VALUES (1) RETURNING id'
 
+    '>> into(table).set(field, 1).with(alias, table)':
+      beforeEach: -> @inst.into('table').set('field', 1).with('alias', squel.select().from('table'))
+      toString: ->
+        assert.same @inst.toString(), 'WITH alias AS (SELECT * FROM table) INSERT INTO table (field) VALUES (1)'
```
Comment (hiddentao) [inline test/postgres.test.coffee]: Please add a `toParam()` test too, and the `with` query should also use params, e.g. `select().from('table').where('a = ?', 5)`

File: src/postgres.js [hunk 4]
```diff
@@ -26,10 +26,67 @@ squel.flavours['postgres'] = function(_squel) {
     }
   }
 
+  // WITH
+  cls.WithBlock = class extends cls.Block {
```
Comment (hiddentao) [inline src/postgres.js]: Please shift this class into the part of the file which contains all the block definitions (it's higher up!).

**Comment by hiddentao** [comment]
Looks good. A few minor points, which I've commented.

**Comment by warrenseymour** [comment]
@hiddentao Thanks. I do have tests for `toParam()` - see here - https://github.com/hiddentao/squel/pull/239/files#diff-aa48ea909f2833a70a5e9b1d6f3e63eaR136 - I was just following the conventions set out by the existing tests, but can easily add parameterized queries for the others too if you're sure.

I'll write up some docs to go with this in the meantime.

**Comment by hiddentao** [comment]
Thanks. I think if you just add some additional parameterized queries for the parts I commented on that would be sufficient.

**Comment by warrenseymour** [comment]
Done, with accompanying docs available in #241

**Comment by hiddentao** [comment]
Thanks, I've commented on #241.

**Pull request merged**